### ISSUE TITLE
AMOEBA vdW softcore support

### ISSUE THREAD

**Issue by mjschnie** [openmm/openmm#2369]
This issue is a discussion of adding AMOEBA Buffered-14-7 softcore support to OpenMM. There are potentially two approaches:

1) Add a "CustomNonbonded" force class that supports reduced hydrogen positions. This would allow hydrogen atom vdW centers to be located between their current nuclear coordinates and the bonded heavy atom.

2) Add support directly to the AmoebaVdwForce. This would entail two global softcore parameters (i.e. power on lambda and a softcore offset), a per-particle Lambda parameter, and  rules for how particles should interact if both have Lambda values < 1.0 (i.e. intramolecular softcore).

The former approach is more general, but may not be widely used since reduced hydrogen positions are relatively uncommon. Approach 2 has been implemented in the TINKER-OpenMM branch:
https://github.com/TinkerTools/Tinker-OpenMM

**Comment by peastman** [comment]
I agree the second approach is probably simpler.  What would the functional form be?  That is, how exactly would each interaction be calculated given the global and per-particle parameters?

**Comment by mjschnie** [comment]
It's Equation 2 of the following paper:
https://www.pnas.org/content/105/17/6290

An initial reference implementation is given in the code linked below. This is a bit more complicated than necessary, because it supports both LJ 6-12 and Buffered-14-7. I'd suggest we only merge softcore Buffered 14-7 support (the LJ 6-12 softcore support is already available using a CustomNonbonded Force).

https://github.com/TinkerTools/Tinker-OpenMM/blob/master/plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceVdwForce.cpp

**Comment by peastman** [comment]
So in that equation, n and p are global parameters, while lambda is a per-particle parameter?  If so, how does the lambda in the equation get determined from the per-particle parameters?  Is there a combining rule for that?

Do you actually need multiple lambdas (turning on and off multiple groups at once)?  Or is it sufficient to have a single lambda, and the per-particle parameters just specify which particles are affected by it?  I ask because modifying a global parameter is a lot cheaper than modifying per-particle parameters.

**Comment by mjschnie** [comment]
The global parameters are n and alpha (p is the separation distance rij normalized by the minimum energy separation).

We could definitely use a single global lambda value, and then a per-particle "flag" to specify if the atom is being soft-cored. Alternatively, the "lambda combining rules" used in TINKER-OpenMM are given below and use a per-particle lambda float value:

RealOpenMM combinedLambda = 1.0;
if (_coupleMethod == Decouple) {
   combinedLambda = (lambdas[ii] == lambdas[jj] ? 1.0f : (lambdas[ii] < lambdas[jj] ? lambdas[ii] : lambdas[jj]));
} else if (_coupleMethod == Annihilate) {
   combinedLambda = (lambdas[ii] < lambdas[jj] ? lambdas[ii] : lambdas[jj]);
}

For the "Decouple" case, intramolecular interactions are maintained, while for the "Annihilate" case they are turned off.

**Comment by peastman** [comment]
Ok, sounds like a single global lambda is fine.  Are both combining rules important to support?

**Comment by mjschnie** [comment]
I think it's worthwhile to support both. "Annihilate" is useful for alchemical changes, while "Decouple" is useful absolute solvation free energy or binding affinity.

**Comment by peastman** [comment]
It looks to me like the code in https://github.com/TinkerTools/Tinker-OpenMM/blob/master/plugins/amoeba/openmmapi/include/openmm/AmoebaVdwForce.h and https://github.com/TinkerTools/Tinker-OpenMM/blob/master/plugins/amoeba/platforms/reference/src/SimTKReference/AmoebaReferenceVdwForce.cpp is actually a different functional form from what you described?  There are no options for setting n or alpha.  And looking at the implementation, the only effect of lambda is to act as an overall scale on the energy.  So what they've implemented is basically equivalent to the equation in the paper with alpha hardcoded to 0?

So it sounds like we want an API along these lines:

1. Add methods for getting and setting n and alpha, as well as the coupling method.  These must be set before the Context is created.  They don't change during a simulation.

2. The force will define a Context parameter called `AmoebaVdwLambda`.  There's no need for a method to set the default value, since of course it will always default to 1.

3. Add a boolean per-particle parameter called `isAlchemical` (or something like that) to specify whether each particle is affected by lambda.

Does that seem about right?

**Comment by mjschnie** [comment]
Indeed, the current implementation in TINKER-OpenMM hard codes n to 5 (lines 149 and 150 in AmoebaReferenceVdwForce.cpp):
RealOpenMM comblambda2 = combinedLambda * combinedLambda;
combinedEpsilon = combinedEpsilon * comblambda2 * comblambda2 * combinedLambda;

And hard codes alpha to 0.7 (line 157):
RealOpenMM scal = 0.7 * (1 - combinedLambda) * (1 - combinedLambda);

For Number 1: the choices of n=5 and alpha=0.7 are good defaults, but it would be great to have getter/setter methods as you mentioned.

Number 2 sounds great; can the AmoebaVdwLambda parameter be updated in an existing Context (i.e. for Lambda Monte Carlo moves)?

Number 3 also sounds perfect.

**Comment by peastman** [comment]
> can the AmoebaVdwLambda parameter be updated in an existing Context (i.e. for Lambda Monte Carlo moves)?

Yes, exactly.  You would call `context.setParameter()`.  That has minimal overhead, in contrast to setting per-particle parameters which can be a bit expensive.

Next question is who is going to implement this. :)  Are you up for doing it?

**Comment by peastman** [comment]
@pren this discussion is relevant to you.  Does the API described above work for your purposes?

**Comment by pren** [comment]
Yes we can read in parameters for the n and a. @leucinw Chengwen can help.

**Comment by mjschnie** [comment]
@peastman Sure - I can work on implementing these additions with help from Pengyu's lab. I'll aim to have a version working in ~1 week or so.

**Comment by peastman** [comment]
Excellent!  Let me know if you have any questions.

**Comment by mjschnie** [comment]
@peastman What do you think of these proof-of-concept changes to the API?

https://github.com/mjschnie/openmm/commit/458f19f3c56b51834bf3f6de72f51d0514f1eb2f

If these Header file changes are consistent with what you're thinking, I'll propagate the implementation on down...

**Comment by peastman** [comment]
Looks good!  On `addParticle()` and `setParticleParameters()` I'd suggest including a default value `isAlchemical=false`.  That helps backward compatibility.  Also, for the global parameter it's good to provide a method that returns the parameter's name.  See, for example, https://github.com/openmm/openmm/blob/master/openmmapi/include/openmm/AndersenThermostat.h#L48-L55.

**Comment by peastman** [comment]
I also noticed you have `getLambda()` and `setLambda()` methods.  Those shouldn't be there.  The value of lambda is stored in the Context, not the force.  You set it with `context.setParameter(AmoebaVdwForce::Lambda(), lambda)`.

**Comment by mjschnie** [comment]
Thanks for the tips. How about this updated version? 
https://github.com/mjschnie/openmm/blob/master/plugins/amoeba/openmmapi/include/openmm/AmoebaVdwForce.h

**Comment by peastman** [comment]
That looks great.

**Comment by mjschnie** [comment]
@peastman Two quick questions.

First, what should the code look like to query the Context for a named parameter (that may not have been set?)? Here's what I have so far:
https://github.com/mjschnie/openmm/blob/36ad9ee7b6390815ef41175721a9dfe8dc2951da/plugins/amoeba/platforms/reference/src/AmoebaReferenceKernels.cpp#L1062

Second, I get the following build error because I'm not sure how to specify the units of the new isAlchemical flag to be "None":

[ 98%] Creating OpenMM Python swig input files...
...
AmoebaVdwForce.getParticleParameters() returns parentIndex as None
AmoebaVdwForce.getParticleParameters() returns sigma as unit.nanometer
AmoebaVdwForce.getParticleParameters() returns epsilon as unit.kilojoule_per_mole
AmoebaVdwForce.getParticleParameters() returns reductionFactor as None
Traceback (most recent call last):
  File "/Data/software/openmm-schnieders-build/python/src/swig_doxygen/swigInputBuilder.py", 
...
Exception: missing unit type for AmoebaVdwForce.getParticleParameters() arg named isAlchemical

**Comment by peastman** [comment]
It looks like you're calling it correctly, though instead of hardcoding `"AmoebaVdwLambda"` it's a little cleaner to write `AmoebaVdwForce::Lambda()`.  All parameters get initialized automatically when the Context is created, so that should always work.  Make sure you implement `getDefaultParameters()` on the ForceImpl to define the names and default values of all parameters.

The units for `getParticleParameters()` are specified at https://github.com/openmm/openmm/blob/master/wrappers/python/src/swig_doxygen/swigInputConfig.py#L320.  It lists the units for each argument, in order.  Since `isAlchemical` is unitless, put `None` for its units.

**Comment by mjschnie** [comment]
Great - thanks Peter! 

As of this Commit: 
https://github.com/mjschnie/openmm/commit/612fc60bd957316ea620e0744d65c66062d18fea

the Reference implementation builds and the following tests pass:
TestReferenceAmoebaVdwForce
TestSerializeAmoebaVdwForce

I'll start on the CUDA changes now.

**Comment by peastman** [comment]
Great!

**Comment by mjschnie** [comment]
Hi Peter @peastman,
Could you take a look at this rough draft CUDA code for the AMOEBA vdW software support:
https://github.com/mjschnie/openmm/blob/eda747cf070a33c2961acfb8c0a31d50859fb448/plugins/amoeba/platforms/cuda/src/kernels/amoebaVdwForce2.cu#L30

Any pointers how to implement CUDA kernel processing to support: 
USE_SOFTCORE
VDW_LAMBDA
VDW_SOFTCORE_POWER
VDW_SOFTCORE_ALPHA

And make available the new per particle flags
isAlchemicalI 
isAlchemicalJ

Thanks!,
- Mike

**Comment by peastman** [comment]
This code:

```c++
bool both = isAlchemicalI && isAlchemicalJ;
bool either = isAlchemicalI || isAlchemicalJ;
bool soft = !both && either;
```

could be simplified to just

```c++
bool soft = (isAlchemicalI != isAlchemicalJ);
```

That code gets executed as part of the standard nonbonded kernel built by CudaNonbondedUtilities.  It gets added at https://github.com/openmm/openmm/blob/master/plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp#L2434-L2435.  Call `addParameter()` on it to add another per-particle parameter, and `addArgument()` to add an argument to the kernel for passing in the current lambda value.  See https://github.com/openmm/openmm/blob/master/plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp#L2401.  For constants like VDW_SOFTCORE_POWER, it's best to compile them into the kernel since that's more efficient than storing them in variables.  Just add them to `replacements` along with all the other constants.

**Comment by mjschnie** [comment]
Thanks Peter!
I think we're getting close. The Amoeba CUDA platform builds now with the vdW softcore changes, but running TestCudaAmoebaVdwForce results in the following error: 

/localscratch/1858020.1.MS/openmmTempKernel0x179fe10_48938.cu(834): error: more than one instance of overloaded function "real_shfl" matches the argument list:
            function "real_shfl(float, int)"
            function "real_shfl(double, int)"
            function "real_shfl(long long, int)"
            argument types are: (int, unsigned int)

Please take a look at 1) the updated Kernel and 2) how alchemical info is being passed in:
1) https://github.com/mjschnie/openmm/blob/cd3e62317daaf88350479234df83bfeee89fe87e/plugins/amoeba/platforms/cuda/src/kernels/amoebaVdwForce2.cu#L29

2) https://github.com/mjschnie/openmm/blob/cd3e62317daaf88350479234df83bfeee89fe87e/plugins/amoeba/platforms/cuda/src/AmoebaCudaKernels.cpp#L2365

Any thoughts would be appreciated!

**Comment by peastman** [comment]
I think I see the problem.  CudaNonbondedUtilities is expecting per-particle parameters to have floating point types.  It wouldn't be hard to make it support integers as well, but in this case you might as well just make `isAlchemical` a float instead of an int.  You're only using it as a flag, so it doesn't much matter whether it's an int that can be 0 or 1, or a float that can be 0.0 or 1.0.

**Comment by mjschnie** [comment]
Thanks Peter -- making the CUDA isAlchemical array a float solved the problem. The updated CUDA AmoebaVdwForce kernel now builds and TestCudaAmoebaVdwForce passes. 

I'll implement some further tests of the new Alchemical API and then inquire about how to go about a pull request.

**Comment by mjschnie** [comment]
The AMOEBA vdW softcore tests are now implemented for the Reference and Cuda platforms. We can probably start to think about a pull request from this commit:
https://github.com/mjschnie/openmm/commit/9c95f25c7cf78f834ff296591b8d716dce838cf1

@peastman What do you think?

**Comment by peastman** [comment]
Sure, go ahead and do a pull request.  It will be easier to review when I can conveniently see exactly what has changed.

**Comment by peastman** [comment]
This was implemented in #2377.
